string start
string end
string[] path